### PR TITLE
RzJson: support simple JSON paths

### LINKED ISSUE #1108: Support JSON Path in `librz/util/json_parser.c`
To ease the access of corresponding key and array number if available, something like `[0].methods[2].flags[0]` where first `[0]` will take first element of the JSON array, then key `"methods"`, which contains an array in turn, then take third element from that array, search for `"flags"` key, which has an array as a value, and take a first element of that array.

Files to change:
- `librz/util/json_parser.c`
- `librz/include/rz_util/rz_json.h`

### PR DESCRIPTION
**Pull request by herrgahr**
add rz_json_get_path, which allows querying by paths like `[0].sub.children[0].thing`

 

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Support simple JSON paths like the one above. Could be extended to support fancy things like `things[*].whatever`, but I don't know if this is needed and what the idiomatic API for this would look like in rizin land.

**Test plan**

Tests were added in `test_json.c`so:
ninja test

**Closing issues**

closes #1108

### CODE REVIEW

File: librz/util/json_parser.c [hunk 1]
```diff
@@ -407,3 +413,55 @@ RZ_API const RzJson *rz_json_item(const RzJson *json, size_t idx) {
...
+			}
+			path = key + keysize;
+			break;
+		default:
+			//TODO report error
+			RZ_JSON_REPORT_ERROR("JSON path: unexpected char", path - 1);
+			return NULL;
+		}
+	}
+	// unreachable code
```
Comment (herrgahr) [inline librz/util/json_parser.c]: Would `rz_warn_if_reached` be the proper thing to use here?

Comment (herrgahr) [reply librz/util/json_parser.c]: Nevermind, I changed the control flow to be more obvious, removing this dead code completely

**Comment by wargio** [comment]
what if we want to iterate all arrays ?

**Comment by herrgahr** [comment]
> what if we want to iterate all arrays ?

do you mean wildcard indices like `things[*].whatever`?

as a first shot, I did the simple thing but this could be done, too.

What would you imagine the API to be like?

For more than one wildcard like `things[*].children[*].whatever` it could be done recursively (with a callback?) or we'd have to allocate&keep state and require the user to clean up afterwards.

Is there any precedence for a similar API in the codebase?

**Comment by XVilka** [comment]
The original issue and use case didn't require wildcards, so it's fine as is, I think.

**Comment by wargio** [comment]
agree

**Review by XVilka — APPROVED** [review]

File: librz/util/json_parser.c [hunk 2]
```diff
@@ -407,3 +413,53 @@ RZ_API const RzJson *rz_json_item(const RzJson *json, size_t idx) {
...
+			++path;
+			js = rz_json_item(js, index);
+			if (!js) {
+				return NULL;
+			}
+			break;
+		case '.':
+			key = path;
+			for (keysize = 0; key[keysize]; ++keysize) {
+				if (strchr(".[", key[keysize]))
```
Comment (wargio) [inline librz/util/json_parser.c]: parenthesis

**Review by wargio — CHANGES_REQUESTED** [review]
LGTM

File: librz/util/json_parser.c (hunk 2 above)
Comment (herrgahr) [reply librz/util/json_parser.c]: done

**Review by XVilka — APPROVED** [review]

**Pull request merged**